MON-3547: Configuration CRD for the Cluster Monitoring Operator (#1627)
openshift/enhancements: enhancements/monitoring/monitoring-config.md

---
title: Configuration CRD for the Cluster Monitoring Operator
authors:
  - "@marioferh"
  - "@danielmellado"
  - "@jan--f"
  - "@moadz"
  - "simonpasquier"
reviewers:
  - "@jan--f"
  - "@moadz"
  - "simonpasquier"
approvers:
  - "@jan--f"
  - "@moadz"
  - "simonpasquier"
  - "@openshift/openshift-team-monitoring"
creation-date: 2024-04-26
last-updated: 2024-04-26
status: provisional
---

# CRD ClusterMonitoring

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

> * Currently, the monitoring stack is configured using a configmaps. In OpenShift though the best practice is to configure operators using custom resources.

Reviewer: ```suggestion
* Currently, the OCP monitoring stacks are configured using ConfigMaps (1 for platform monitoring and 1 for user-defined monitoring). In OpenShift though the best practice is to configure operators using custom resources.
```

## Motivation

> * The specification is well known and to a degree self-documenting

Reviewer: ```suggestion
* The specification is well known and to a degree self-documenting.
```

> * The APIServer will validate user resources based on our specifications, so users get immediate feedback on errors instead of having to check if their config was applied and check logs.

Reviewer: ```suggestion
* The Kubernetes API Server validate custom resources based on their API specification, so users get immediate feedback on errors instead of checking the ClusterOperator object and the cluster monitoring operator's logs. Custom expression language (CEL) even allows to write complex validation logic, including cross-field tests.
```

> * Many users expect to interact with operators through a CRD

Reviewer: (nit)
```suggestion
* Many users expect to interact with operators through a CRD.
```

* Compatible with GitOps workflows. 

> * We can add [cross]validation rules to CRD fields to avoid misconfigurations

Reviewer: It looks redundant to me with the 2nd bullet point.

Author: right, sorry c&p

> * End users get a much faster feedback loop. No more applying the config and scanning logs if things don't look right. The API server will give immediate feedback

Reviewer: Same as 2nd point too?

Author: right, sorry c&p

> * Organizational users (such as ACM) can manage a single resource and observe its status

Reviewer: We can link to https://issues.redhat.com/browse/ACM-2545. It can also be merged with the GitOps argument since it's kind of the same request.
```suggestion
* CRDs supports multiple actors managing the same resource which is a key property for the Observability service of Advanced Cluster Management.
```

### Goals

- Replace configmaps with CRD
- Smooth transition for users

## Proposal

### Overview

> - Replace confimgaps with CRD:

Reviewer: Can you add a bit more details on the current implementation and the overall context for readers not familiar with the product?

* Describe the 2 stacks (platform & UWM)
* How does an admin configure each stack? Relationship between the 2?
* Explain that there can be 2 personas responsible for each stack (the managed OpenShift use case). 

A link to https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/multi-tenant-alerting.md would be useful.

Author: Add improvements

  cluster-monitoring-configmap
  user-workload-monitoring-config

  pkg/manifests/config.go

```
  type Config struct {
	Images                               *Images `json:"-"`
	RemoteWrite                          bool    `json:"-"`
	CollectionProfilesFeatureGateEnabled bool    `json:"-"`

	ClusterMonitoringConfiguration *ClusterMonitoringConfiguration `json:"-"`
	UserWorkloadConfiguration      *UserWorkloadConfiguration      `json:"-"`
}
```

> 

Author: Add improvements

### Migration path

- Feature gate. 
- Move each component to openshift/api/config

> - Create two CRD's:

Reviewer: This should go in the previous section IMHO.

Author: done

- Review/improve/modify every configmap component to create a compliant openshif/api type. Fix erros, clean old things, improve names and config.
- Switch mecanishm
- Both ConfigMap and CRD will coexist for a while until the transition is complete.
- CRD and ConfigMap will be merged into a single structure, this structure will config CMO
- Extensive testing will be required to validate the new behavior.
- If there are different fields in CRD and ConfigMap, the field in the ConfigMap will take precedence.
- Some fields in CRD and ConfigMap diverge due to API requirements. Helper functions will be needed to resolve this.
- The end user will be able to choose to use ConfigMap or CRD at any time, keeping in mind that the fields in the CRD will gradually expand.

### Issues

- Merge CRD and configmap could be error prone and will need extra test
- Change of types and names and how handle it when there are CRD and configmap

> 

Reviewer: should these items go to "Open Questions" for now?

Author: done

---
title: Configuration CRD for the Cluster Monitoring Operator
authors:
  - "@marioferh"
  - "@danielmellado"
  - "@jan--f"
  - "@moadz"

>   - "simonpasquier"

Reviewer 2: ```suggestion
  - "@simonpasquier"
```

reviewers:
  - "@jan--f"
  - "@moadz"

>   - "simonpasquier"

Reviewer 2: ```suggestion
  - "@simonpasquier"
```

approvers:
  - "@jan--f"
  - "@moadz"

>   - "simonpasquier"

Reviewer 2: ```suggestion
  - "@simonpasquier"
```

  - "@openshift/openshift-team-monitoring"
creation-date: 2024-04-26
last-updated: 2024-04-26
status: provisional
---

# CRD ClusterMonitoring

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

* Currently, the OCP monitoring stacks are configured using ConfigMaps (1 for platform monitoring and 1 for user-defined monitoring). In OpenShift though the best practice is to configure operators using custom resources.

## Motivation

* The specification is well known and to a degree self-documenting.

> * The Kubernetes API Server validate custom resources based on their API specification, so users get immediate feedback on errors instead of checking the ClusterOperator object and the cluster monitoring operator's logs. Custom expression language (CEL) even allows to write complex validation logic, including cross-field tests.

Reviewer 3: Nit: Common expressions language, not Custom expressions language

> * The Kubernetes API Server validate custom resources based on their API specification, so users get immediate feedback on errors instead of checking the ClusterOperator object and the cluster monitoring operator's logs. Custom expression language (CEL) even allows to write complex validation logic, including cross-field tests.

Reviewer 2: ```suggestion
* The Kubernetes API Server validates custom resources based on their API specification, so users get immediate feedback on errors instead of checking the ClusterOperator object and the cluster monitoring operator's logs. Custom expression language (CEL) even allows to write complex validation logic, including cross-field tests.
```

* Many users expect to interact with operators through a CRD.
* Custom resources play better with GitOps workflows. 
* CRDs supports multiple actors managing the same resource which is a key property for the Observability service of Advanced Cluster Management.

### Goals

- Replace the existing ConfigMaps with CRDs.
- Automated and friction-less upgrade for users.

## Proposal

### Overview

> Currently in CMO a config map provides a way to inject configuration data into pods. There are two configmaps for the different stacks:

Reviewer 2: ```suggestion
Currently in CMO, a config map provides a way to inject configuration data into pods. There are two configmaps for the different stacks:
```

>     cluster-monitoring-configmap: Default platform monitoring components. A set of platform monitoring components are installed in the openshift-monitoring project by default during an OpenShift Container Platform installation. This provides monitoring for core cluster components including Kubernetes services. The default monitoring stack also enables remote health monitoring for clusters.

Reviewer 2: ```suggestion
    cluster-monitoring-config: Default platform monitoring components. A set of platform monitoring components are installed in the openshift-monitoring project by default during an OpenShift Container Platform installation. This provides monitoring for core cluster components including Kubernetes services. The default monitoring stack also enables remote health monitoring for clusters.
```

Two distinct CRDs are necessary because they are managed by different personas with specific roles and responsibilities:

    - UWM admins: manage the configuration of the UWM components (edit permissions on the openshift-user-workload-monitoring/user-workload-monitoring-config configmap).
    - Cluster admins: manage the configuration of the Platform monitoring components.

In managed OpenShift clusters like OSD/ROSA, two separate CRDs are necessary because platform SREs manage the cluster's platform monitoring stack, while customers manage the user-defined monitoring stack. This separation ensures that each group maintains control over their specific monitoring configurations, reducing conflicts and enhancing system management.

[More info](https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/multi-tenant-alerting.md)

> - Replace confimgaps with CRD:

Reviewer 2: ```suggestion
- Replace configmaps with CRD:
```

  
```

>   type Config struct {

Reviewer 3: Is this the existing config?

	Images                               *Images `json:"-"`
	RemoteWrite                          bool    `json:"-"`
	CollectionProfilesFeatureGateEnabled bool    `json:"-"`

	ClusterMonitoringConfiguration *ClusterMonitoringConfiguration `json:"-"`
	UserWorkloadConfiguration      *UserWorkloadConfiguration      `json:"-"`
}
```

We will strive to maintain the previous structure as much as possible while adapting it to OpenShift API conventions. 

```

> type ClusterMonitoringConfiguration struct {

Reviewer 3: Is this the existing state? Or desired?

	KubeStateMetricsConfig *KubeStateMetricsConfig `json:"kubeStateMetrics,omitempty"`
	PrometheusK8sConfig *PrometheusK8sConfig `json:"prometheusK8s,omitempty"`
	PrometheusOperatorConfig *PrometheusOperatorConfig `json:"prometheusOperator,omitempty"`
	PrometheusOperatorAdmissionWebhookConfig *PrometheusOperatorAdmissionWebhookConfig `json:"prometheusOperatorAdmissionWebhook,omitempty"`
	OpenShiftMetricsConfig *OpenShiftStateMetricsConfig `json:"openshiftStateMetrics,omitempty"`
	TelemeterClientConfig *TelemeterClientConfig `json:"telemeterClient,omitempty"`
	ThanosQuerierConfig *ThanosQuerierConfig `json:"thanosQuerier,omitempty"`
	NodeExporterConfig NodeExporterConfig `json:"nodeExporter,omitempty"`
	MonitoringPluginConfig *MonitoringPluginConfig `json:"monitoringPlugin,omitempty"`
}
```

Each component within the ConfigMap will be migrated to the OpenShift API in separate PRs. This approach allows for a thorough review, improvement, and modification of each ConfigMap component to ensure it aligns with OpenShift API standards. As part of this process, types will be modified, outdated elements will be removed and names and configurations will be refined.

### Migration path

> - Switch mecanishm

Reviewer 2: ```suggestion
- Switch mechanism
```

> - Both ConfigMap and CRD will coexist for a while until the transition is complete.

Reviewer 3: How will clusters that do not currently have a CR configured, but do have the configmap, continue to function without manual intervention as you go through upgrades?

Author: "The feature will be controlled by a feature gate. However, if a user wants to test it, the CR and ConfigMap will be merged into a single structure, making the process seamless and transparent for the user."

- CRD and ConfigMap will be merged into a single structure, this structure will config CMO
- Extensive testing will be required to validate the new behavior.
- If there are different fields in CRD and ConfigMap, the field in the ConfigMap will take precedence.
- Some fields in CRD and ConfigMap diverge due to API requirements. Helper functions will be needed to resolve this.

> - The end user will be able to choose to use ConfigMap or CRD at any time, keeping in mind that the fields in the CRD will gradually expand.

Reviewer 3: Will this be a conscious choice? Is there some explicit switch for this or is it just a case of if one or the other exists?

Reviewer 2: does user still can choose to use ConfigMap or CRD after the feature GA?

Author: The feature gate acts as the switch for this functionality, which is a deliberate choice since the CR will continue to expand until it is fully developed. After the feature reaches GA, users can choose between the CR and ConfigMap, unless there is a strong consensus to remove the ConfigMap

### Open questions

- Merge CRD and configmap could be error prone and will need extra test
- Change of types and names and how handle it when there are CRD and configmap

### Transition to the user

- How the user could adopt CRD instead of configmap.

## Design Details

> To initiate the process, let's establish a feature gate that will serve as the entry point for implementing a CRD configuration approach. This strategy enables us to make incremental advancements without the immediate burden of achieving complete feature equivalence with the config map. We can commence with a the basics and progressively incorporate additional functionalities as they develop.

Reviewer 2: ```suggestion
To initiate the process, let's establish a feature gate that will serve as the entry point for implementing a CRD configuration approach. This strategy enables us to make incremental advancements without the immediate burden of achieving complete feature equivalence with the config map. We can commence with the basics and progressively incorporate additional functionalities as they develop.
```

One proposal for a minimal DoD was:

>     - Feature gate in openshift/api

Reviewer 2: ```suggestion
- Feature gate in openshift/api
```
it seems you want to get a list, but the format is 
```
One proposal for a minimal DoD was: - Feature gate in openshift/api - Api types moved to openshift/api - CRD Initial dev https://github.com/openshift/api/pull/1929 Add controller-gen logic to makefile Add API to api/config/v1 Add Generated CRD: config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_clustermonitoring.crd.yaml Add example CustomResource: config/examples/clustermonitoringoperator.yaml - Client codegen - Reconcile logic: https://github.com/openshift/cluster-monitoring-operator/pull/2350 - Add decoupling Confimgap / CustomResource: Controller logic is strongly dependant of *manifests.Config struct.
```

>     - Api types moved to openshift/api

Reviewer 2: ```suggestion
    - API types moved to openshift/api
```

    - CRD Initial dev https://github.com/openshift/api/pull/1929
        Add controller-gen logic to makefile
        Add API to api/config/v1
        Add Generated CRD: config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_clustermonitoring.crd.yaml
        Add example CustomResource: config/examples/clustermonitoringoperator.yaml
    - Client codegen 
    - Reconcile logic: https://github.com/openshift/cluster-monitoring-operator/pull/2350

>     - Add decoupling Confimgap / CustomResource:

Reviewer 2: ```suggestion
    - Add decoupling Configmap / CustomResource:
```

        Controller logic is strongly dependant of *manifests.Config struct.
        

### Example configuration

#### current configmap

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    telemeterClient:
      enabled: false
    prometheusK8s:
      volumeClaimTemplate:
        metadata:
          name: prometheus-data
          annotations:
            openshift.io/cluster-monitoring-drop-pvc: "yes"
        spec:
          resources:
            requests:
              storage: 20Gi

> 

Reviewer 2: could remove the empty line

```

### CRD

```
apiVersion: clustermonitoring.config.openshift.io
kind: ClusterMonitoring
metadata:
  name: cluster
  namespace: openshift-monitoring
spec:
  prometheusK8s:
    volumeClaimTemplate:
      metadata:
        name: prometheus-data
        annotations:
          openshift.io/cluster-monitoring-drop-pvc: "yes"
      spec:
        resources:
          requests:

>             storage: 20G`

Reviewer 2: ```suggestion
            storage: 20G
```
```suggestion
            storage: 20Gi
```

```

### Test Plan

- Unit tests for the feature

> - e2e tests covering the feature

Reviewer 2: I think upgrade testing should be covered

---

> title: monitoring-config 

Reviewer 4: this has an additional trailing space after monitoring-config

Author: done

        Controller logic is strongly dependant of *manifests.Config struct.
        

### Example configuration

#### current ConfigMap

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    telemeterClient:
      enabled: false
    prometheusK8s:
      volumeClaimTemplate:
        metadata:
          name: prometheus-data
          annotations:
            openshift.io/cluster-monitoring-drop-pvc: "yes"
        spec:
          resources:
            requests:
              storage: 20Gi
```

### CRD

```
apiVersion: clustermonitoring.config.openshift.io
kind: ClusterMonitoring
metadata:
  name: cluster
  namespace: openshift-monitoring
spec:
  prometheusK8s:
    volumeClaimTemplate:
      metadata:
        name: prometheus-data
        annotations:
          openshift.io/cluster-monitoring-drop-pvc: "yes"
      spec:
        resources:
          requests:
            storage: 20G
```

### Workflow Description

See example in the [Proposal](#Proposal) for details.

### API Extensions

Add two new CRD's:
- ClusterMonitoring
- UserWorkloadClusterMonitoring

### Migration path

> - Switch mechanism

Reviewer 4: In terms of the migration path, I'd elaborate this a bit further and explain possible cornercases. How do you migrate a field from CM to CRD? just as an example.

- Both ConfigMap and CRD will coexist for a while until the transition is complete.
- CRD and ConfigMap will be merged into a single structure, this structure will config CMO
- Extensive testing will be required to validate the new behavior.
- If there are different fields in CRD and ConfigMap, the field in the ConfigMap will take precedence.
- Some fields in CRD and ConfigMap diverge due to API requirements. Helper functions will be needed to resolve this.
- The end user will be able to choose to use ConfigMap or CRD at any time, keeping in mind that the fields in the CRD will gradually expand.

### Risks and Mitigations

> - Issues when ConfigMap and CR coexist

Reviewer 4: I see that you address that partially here, I'm fine having it here but will extend the explanation a bit.

Author: done

- Mismatch in the configuration of operators with ConfigMap and CR

### Drawbacks

> - Migration to user

Reviewer 4: What do you mean with 'Migration to user'?

Author: done

- Coexisting of both ConfigMap and CRD

### Test Plan

- Unit tests for the feature
- e2e tests covering the feature
- Upgrade testing

> 

Reviewer 4: I'd also put an item for the FeatureGate

Author: done

	AlertmanagerMainConfig *AlertmanagerMainConfig `json:"alertmanagerMain,omitempty"`
	UserWorkloadEnabled *bool `json:"enableUserWorkload,omitempty"`
	HTTPConfig *HTTPConfig `json:"http,omitempty"`
	K8sPrometheusAdapter *K8sPrometheusAdapter `json:"k8sPrometheusAdapter,omitempty"`
	MetricsServerConfig *MetricsServerConfig `json:"metricsServer,omitempty"`
	KubeStateMetricsConfig *KubeStateMetricsConfig `json:"kubeStateMetrics,omitempty"`
	PrometheusK8sConfig *PrometheusK8sConfig `json:"prometheusK8s,omitempty"`
	PrometheusOperatorConfig *PrometheusOperatorConfig `json:"prometheusOperator,omitempty"`
	PrometheusOperatorAdmissionWebhookConfig *PrometheusOperatorAdmissionWebhookConfig `json:"prometheusOperatorAdmissionWebhook,omitempty"`
	OpenShiftMetricsConfig *OpenShiftStateMetricsConfig `json:"openshiftStateMetrics,omitempty"`
	TelemeterClientConfig *TelemeterClientConfig `json:"telemeterClient,omitempty"`
	ThanosQuerierConfig *ThanosQuerierConfig `json:"thanosQuerier,omitempty"`
	NodeExporterConfig NodeExporterConfig `json:"nodeExporter,omitempty"`
	MonitoringPluginConfig *MonitoringPluginConfig `json:"monitoringPlugin,omitempty"`
}
```

> Each component within the ConfigMap will be migrated to the OpenShift API in separate PRs. This approach allows for a thorough review, improvement, and modification of each ConfigMap component to ensure it aligns with OpenShift API standards. As part of this process, types will be modified, outdated elements will be removed and names and configurations will be refined.

Reviewer 5: Yeah that is the plan. Should this be mentioned here or are you looking for a more detailed plan?

> Each component within the ConfigMap will be migrated to the OpenShift API in separate PRs. This approach allows for a thorough review, improvement, and modification of each ConfigMap component to ensure it aligns with OpenShift API standards. As part of this process, types will be modified, outdated elements will be removed and names and configurations will be refined.

Reviewer 3: I would certainly mention it here, and, perhaps if possible, create a list of the ones you know are co-dependent and therefore should be considered together

> Each component within the ConfigMap will be migrated to the OpenShift API in separate PRs. This approach allows for a thorough review, improvement, and modification of each ConfigMap component to ensure it aligns with OpenShift API standards. As part of this process, types will be modified, outdated elements will be removed and names and configurations will be refined.

Author: There are a lot components, not sure all the co-dependencies. The idea was to migrate one by one, and discuss in every PR the names, types, definitions and possible removes of types or refactors. After that we can think in handle the co-dependence. 

At the very beginning we tried to migrate all at the same time to see the similar fields and dependencies but its a huge task and no so useful.

> Each component within the ConfigMap will be migrated to the OpenShift API in separate PRs. This approach allows for a thorough review, improvement, and modification of each ConfigMap component to ensure it aligns with OpenShift API standards. As part of this process, types will be modified, outdated elements will be removed and names and configurations will be refined.

Reviewer 5: Right, our current assumption is that the configmap structure already separates the components pretty well. The only interdependency we have is that the `UserDefinedMonitoring` object will only have an effect if `ClusterMonitoringSpec.UserDefined != Disabled`, speaking in CRD terms. There is a similar interdependency for dedicated alertmanagers for user defined monitoring. It's probably worth calling out those two (soft) dependencies and our assumption that otherwise things are separated.

          annotations:
            openshift.io/cluster-monitoring-drop-pvc: "yes"
        spec:
          resources:
            requests:
              storage: 20Gi
```

### CRD

```
apiVersion: clustermonitoring.config.openshift.io
kind: ClusterMonitoring
metadata:
  name: cluster
  namespace: openshift-monitoring
spec:
  prometheusK8s:
    volumeClaimTemplate:
      metadata:
        name: prometheus-data
        annotations:
          openshift.io/cluster-monitoring-drop-pvc: "yes"
      spec:
        resources:
          requests:
            storage: 20G
```

### Workflow Description

See example in the [Proposal](#Proposal) for details.

### API Extensions

Add two new CRD's:
- ClusterMonitoring
- UserWorkloadClusterMonitoring

### Topology Considerations

Will function identically in all topologies.

#### Hypershift / Hosted Control Planes
No special considerations.

#### Standalone Clusters
No special considerations.

#### Single-node Deployments or MicroShift
No special considerations.

### Implementation Details/Notes/Constraints

### Migration path

- Switch mechanism
- Both ConfigMap and CRD will coexist for a while until the transition is complete.

> - CRD and ConfigMap will be merged into a single structure, this structure will config CMO

Author: done.

- Extensive testing will be required to validate the new behavior.
- If there are different fields in CRD and ConfigMap, the field in the ConfigMap will take precedence.
- Some fields in CRD and ConfigMap diverge due to API requirements. Helper functions will be needed to resolve this.
- The end user will be able to choose to use ConfigMap or CRD at any time, keeping in mind that the fields in the CRD will gradually expand.

### Risks and Mitigations

- Issues when ConfigMap and CRD coexist, as ConfigMap and CRD will be merged in a single structure, this could be error prone and an extensive testing should be done.
- Mismatch in the configuration of operators with ConfigMap and CR

### Drawbacks

- Migration. The user can use either ConfigMap or CRD, or both; we need to explain how to make the transition, and the preference is to use the CRD.
- Coexisting of both ConfigMap and CRD.

## Test Plan

- Unit tests for the feature
- e2e tests covering the feature
- Upgrade testing
- FeatureGate

> ### Open questions

Author: We need to move forward with the development to answer some of them.

- Merge CRD and ConfigMap could be error prone and will need extra test
- Change of types and names and how handle it when there are CRD and ConfigMap
- Which tools/mechanisms do we provide to cluster admins to minimize the migration burden? Should CMO automatically create the custom resource from the current ConfigMap when the CRD becomes GA? Or do we want a semi-automated approach with explicit approval from the cluster admins?
- Should we provide an escape hatch (via some annotation?) telling CMO to ignore the custom resource and only use the ConfigMap in case something goes bad?
- How do we tell cluster admins that they need to migrate? CMO conditions could be an option as well as alerting and Insights?- How do we know how many clusters have migrated to the new CRD? Telemetry I guess.

> - How long should we keep the support for the ConfigMap?

Author: I don't think we are force user to upgrade. It will be a new feature that they can use once is integrated and added to de docs.

> - How long should we keep the support for the ConfigMap?

Reviewer 5: Hmm eventually we will want to force users to migrate. How long we'll support eh config map depends on how quickly we can implement the respective config map fields.
I think we will eventually deprecate the use of the config map. Once we have followed proper deprecation procedures and such we could perhaps block updates when the config map is still present?

> - How long should we keep the support for the ConfigMap?

Author: Sorry, I misspoke. I meant no enforcement until the feature gate is removed. Of course, at some point, we'll have to require users to use only CR, but as you say, that doesn’t need to be addressed here, right?
- Coexist under feature gate
- Remove feature gate when feature works correctly
- start deprecation procedure

> - How long should we keep the support for the ConfigMap?

Reviewer 3: It doesn't need to be done for the initial version no, but might be good to have some ideas down about how you might go about removing it.

> - How long should we keep the support for the ConfigMap?

Author: Some context added. thx @simonpasquier

    - Cluster admins: manage the configuration of the Platform monitoring components.

In managed OpenShift clusters like OSD/ROSA, two separate CRDs are necessary because platform SREs manage the cluster's platform monitoring stack, while customers manage the user-defined monitoring stack. This separation ensures that each group maintains control over their specific monitoring configurations, reducing conflicts and enhancing system management.

[More info](https://github.com/openshift/enhancements/blob/master/enhancements/monitoring/multi-tenant-alerting.md)

- Replace configmaps with CRD:

Current Config struct in CMO
  
```
  type Config struct {
	Images                               *Images `json:"-"`
	RemoteWrite                          bool    `json:"-"`
	CollectionProfilesFeatureGateEnabled bool    `json:"-"`

	ClusterMonitoringConfiguration *ClusterMonitoringConfiguration `json:"-"`
	UserWorkloadConfiguration      *UserWorkloadConfiguration      `json:"-"`
}
```

We will strive to maintain the previous structure as much as possible while adapting it to OpenShift API standards. 

> Current ClusterMonitoringConfiguration struct:

Reviewer 6: can we include a perma link here?

Author: https://github.com/openshift/cluster-monitoring-operator/blob/eca25a9571d721a90f1b7f8d26931c679ae93aec/pkg/manifests/types.go#L36

- Feature gate in openshift/api
    - API types moved to openshift/api
    - CRD Initial dev https://github.com/openshift/api/pull/1929
        Add controller-gen logic to makefile
        Add API to api/config/v1
        Add Generated CRD: config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_clustermonitoring.crd.yaml
        Add example CustomResource: config/examples/clustermonitoringoperator.yaml
    - Client codegen 
    - Reconcile logic: https://github.com/openshift/cluster-monitoring-operator/pull/2350
    - Add decoupling ConfigMap / CustomResource:
        Controller logic is strongly dependant of *manifests.Config struct.
        

### Example configuration

#### current ConfigMap

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    telemeterClient:
      enabled: false
    prometheusK8s:
      volumeClaimTemplate:
        metadata:
          name: prometheus-data
          annotations:
            openshift.io/cluster-monitoring-drop-pvc: "yes"
        spec:
          resources:
            requests:
              storage: 20Gi
```

### CRD

```

> apiVersion: clustermonitoring.config.openshift.io

Reviewer 6: How does this API relates to https://github.com/rhobs/observability-operator/blob/main/docs/api.md
https://github.com/rhobs/observability-operator/blob/main/docs/api.md#monitoringstackspec

Author: https://github.com/openshift/cluster-monitoring-operator/blob/eca25a9571d721a90f1b7f8d26931c679ae93aec/pkg/manifests/types.go#L36

Author: They are not related.

  - The merge process look at top-level fields
  - Behavior: If an entry exists in the ConfigMap, ignore the CRD.
      If there’s no entry in the ConfigMap, use the CRD.
  - Merge Process (only for phase 1)
      Top-level fields (e.g., AlertmanagerMainConfig *AlertmanagerMainConfig || spec.AlertmanagerMainConfig):
      If a field is nil in the ConfigMap, use the CRD; otherwise, stick with the ConfigMap.

- Phase 2: Transition (User Control)
  Discuss whether to keep dual support for a couple of releases.
  - New Logic:
    Allow users to choose the source of configuration (ConfigMap or CRD, warn about CRD preference).
    If a CRD exists, prioritize it over the ConfigMap.
    Add logic for users to manually switch to CRD if desired.
    No merge process.

- Phase 3: Full Migration (CRD as Default)
    After a few releases, the migration from ConfigMap to CRD will be complete.
    ConfigMap will be deprecated, with CRD becoming the default configuration source.

- Extensive testing will be required to validate the new behavior.
- Some fields in CRD and ConfigMap diverge due to API requirements. Helper functions will be needed to resolve this.

> ### Risks and Mitigations

Reviewer 6: The migration path needs to be acked by SREs and other owners of this policy cc @bng0y @typeid 
https://github.com/openshift/hypershift-dataplane-metrics-forwarder/blob/main/hack/templates/00-hypershift-dataplane-metrics-forwarder.SelectorSyncSet.yaml.tmpl#L107

Reviewer 7: Thanks for the ping. This enhancement as a whole as well as the migration plan make sense to me.

---
title: monitoring-config
authors:
  - "@marioferh"
  - "@danielmellado"
  - "@jan--f"
  - "@moadz"
  - "@simonpasquier"
reviewers:
  - "@jan--f"
  - "@moadz"
  - "@simonpasquier"
approvers:
  - "@jan--f"
  - "@moadz"
  - "@simonpasquier"
  - "@openshift/openshift-team-monitoring"
api-approvers:
  - "@deads2k"
  - "@JoelSpeed"
creation-date: 2024-04-26
last-updated: 20204-26
tracking-link:
- "https://issues.redhat.com/browse/MON-1100"
  
---

# CRD ClusterMonitoring

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA
- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)

## Summary

* Currently, the OCP monitoring stacks are configured using ConfigMaps (1 for platform monitoring and 1 for user-defined monitoring). In OpenShift though the best practice is to configure operators using custom resources.

## Motivation

> * The specification is well known and to a degree self-documenting.

Reviewer 3: With proper godoc writing, "to a degree" could be dropped

Author: done

* The Kubernetes API Server validates custom resources based on their API specification, so users get immediate feedback on errors instead of checking the ClusterOperator object and the cluster monitoring operator's logs. Common expressions language even allows to write complex validation logic, including cross-field tests.
* Many users expect to interact with operators through a CRD.
* Custom resources play better with GitOps workflows. 
* CRDs supports multiple actors managing the same resource which is a key property for the Observability service of Advanced Cluster Management.

### User Stories

> As a cluster administrator, I want to be able to add a CRD so that I can configure monitoring stack.

Reviewer 3: IMO we can improve on the user story here. The end user doesn't really care about it being a CRD do they, what they want is a well structured and documented API with feedback when they misconfigure it, so that it is easy for them to configure the monitoring stack. The implementation of that I don't think they really care about.

Author: I'm not sure about it. This proposal focuses on the migration from ConfigMap to CR. We can add another field to explain its functionality or consider a separate proposal, the aim of this one is the migration. I believe it's important for the end user to understand what was possible before this feature and what can be achieved afterward

Reviewer 3: The user story I don't think represents the reasons you're performing this migration. Something has spurred you to do this, feedback from users or something? I suspect the feedback is along the lines of my first comment in this thread

Users can apply configmaps, but there's no documented structure, nor is there any validation. By moving to a CRD, we get those things, but its those user facing improvements that the user story should focus on, not the technology

Author: I can add that to the User Story, but the reasons you wrote here, why we are migrating to CR are in the motivations. We can explain again what monitoring stack does but not sure if is the goal of this proposal

## Motivation

* The specification is well known and largely self-documenting.
* The Kubernetes API Server validates custom resources based on their API specification, so users get immediate feedback on errors instead of checking the ClusterOperator object and the cluster monitoring operator's logs. Common expressions language even allows to write complex validation logic, including cross-field tests.
* Many users expect to interact with operators through a CRD.
* Custom resources play better with GitOps workflows. 
* CRDs supports multiple actors managing the same resource which is a key property for the Observability service of Advanced Cluster Management.

Reviewer 6: I agree with Joel, using a CRD is an impl outcome (the "what"), It'd be good to flesh out user stories driven by concrete value added for different personas (the "why we are investing in this"), i.e can you share a concrete example of how any of the statements in the "motivation" section impacts the cluster admin when setting up a monitoring stack?
e.g.
As a Service Provider I want to have granular ownership over some configuration knobs like remoteWrite so I can forward data to a centralized managed service without other entities clobbering my config.
As a Service Provider I want my managed (bad) configuration to fail on admission so I can propagate and signal the failure appropriately to watching components of the service and to a Service Consumer.

Author: Thanks, I need to think a little bit about what to add in user stories, thanks for the feedback.

Reviewer 5: We had the request to improve a gitops style management for monitoring from various sides. Maybe that is worth mentioning here. Admins want to create a CR (possibly via gitops style systems) and observe the status of the created CR. Today this is spread over the config map and feedback via CMO status conditions and CMO logs.

  prometheusK8s:
    volumeClaimTemplate:
      metadata:
        name: prometheus-data
        annotations:
          openshift.io/cluster-monitoring-drop-pvc: "yes"
      spec:
        resources:
          requests:
            storage: 20G
```

### Workflow Description

See example in the [Proposal](#Proposal) for details.

### API Extensions

Add two new CRD's:
- ClusterMonitoring
- UserWorkloadClusterMonitoring

### Topology Considerations

Will function identically in all topologies.

#### Hypershift / Hosted Control Planes
No special considerations.

#### Standalone Clusters
No special considerations.

#### Single-node Deployments or MicroShift
No special considerations.

### Implementation Details/Notes/Constraints

### Migration path

Migration path it will be split in three phases

- Phase 1: Pre-GA (Gradual Migration for Developers)
  - In this phase only developers are expected to use CRD.
  - Both ConfigMap and CRD will coexist for a while until the transition is complete.
  - Work with a Feature flag.
  - Two options: ConfigMap or CRD.
      Preference is to use ConfigMap over CRD during this phase.
  - Behavior: If an entry exists in the ConfigMap, ignore the CRD.

>       If there’s no entry in the ConfigMap, use the CRD.

Reviewer 3: Please clarify, do you mean for specific fields or the entire CRD/ConfigMap

Author: Yep sorry, moving lines I've dropped this one:
  - The merge process look at top-level fields

  - Merge Process (only for phase 1)
      Top-level fields (e.g., AlertmanagerMainConfig *AlertmanagerMainConfig || spec.AlertmanagerMainConfig):
      If a field is nil in the ConfigMap, use the CRD; otherwise, stick with the ConfigMap.

- Phase 2: Transition (User Control)

>   Discuss whether to keep dual support for a couple of releases.

Reviewer 3: If you do not have dual support, any customer who is currently writing to the configmap (think maybe a GitOps user?) will be broken on upgrade.

You'll need some period of overlap for them to be able to migrate across. 2 releases IMO is the very minimum

Author: Right

Reviewer 3: When is this going to be decided?

Author: We will maintain dual support for at least 2 releases. 
Add to PR

  - New Logic:

>     Allow users to choose the source of configuration (ConfigMap or CRD, warn about CRD preference).

Reviewer 3: By having a config somewhere, or just the presence?

Author: We can warn the user it is possible to use both, but always CR preference, but next point is to add some logic to do it.

Reviewer 3: Can you explain in the doc how you'll do that? This should be represented in the workflows. A workflow that shows what a user would see when they apply the CR or configmap in this dual state would really help

Author: I am not sure to understand it. If user choose a CR, logic take CR and monitoring stack is configured. If user choose ConfigMap, then there is no CR and logic take ConfigMap. If the is the case we have both, we use CR preference.

Reviewer 3: So if the user creates the CR, then a different user on the same cluster updates the configmap, expecting the change to take effect, they won't see any change since it will be overridden by the CR.

How would that user realise that they are configuring the wrong object? Can we build something (ValidatingAdmissionPolicy) that would warn them if they update the configmap when the CR exists?

Author: Yep it can be a hook. We can choose the best option to warn the user.

Reviewer 3: I don't mean a hook, I mean ValidatingAdmissionPolicy specifically, it's more performant and easier to write than a webhook

Author: Add new info, please close discussion or add suggestion in order to clean the PR. Thanks!

    If a CRD exists, prioritize it over the ConfigMap.

>     Add logic for users to manually switch to CRD if desired.

Reviewer 3: What would the workflow look like for this? Will you build a tool that converts the configmap to CRD for them?

Author: at this point we dont have a merge process, we should choose between CR or configmap.

Author: We can add some tips to do the migration, but  user should read CR API and configure every type. We will work to improve fields as in

` This approach allows for a thorough review, improvement, and modification of each ConfigMap component to ensure it aligns with OpenShift API standards. As part of this process, types will be modified, outdated elements will be removed and names and configurations will be refined.`

Then users should read the release notes, the API, the tips or whatever and create a new CR.

Reviewer 3: > We can add some tips to do the migration, but user should read CR API and configure every type.

You will have lazy users, who do not want to do this, and therefore will stay on the configmap forever. Is there anything we can do to provide a tool that would convert the configuration for them, as that possibly unblock you being able to remove configmap support

Author: I don't think is a good idea, CR is the standard way to defined resources and users are habituated to deal with it. To add an extra tool to will add extra work and could be error prone with different versions/releases.

Reviewer 3: Maybe not an extra tool then, but you will need to provide a migration path before you can remove support for the configmap.

That may well just be documentation, but you'll likely need the end of that documentation to result in deleting the configmap, so that you can, in the future, block upgrades for anyone who has not yet migrated.

I would also consider folks like service delivery, do they use the configmap today? What would migration look like for them, can they easily script the migration?

Author: Add new info, please close discussion or add suggestion in order to clean the PR. Thanks!

    No merge process.

- Phase 3: Full Migration (CRD as Default)
    After a few releases, the migration from ConfigMap to CRD will be complete.

>     ConfigMap will be deprecated, with CRD becoming the default configuration source.

Reviewer 3: In this mode, the configmap is no longer read from at all? Is that correct?

I think technically you need deprecation to be earlier than the removal, so deprecation would likely be phase 2, removal in phase 3

Author: If is in deprecation mode, it can be read but not recommended to do it.
Our idea is to do deprecation if phase 3, but we can also do it in phase 2. Something like in phase 2 we want to try it but not force user to do it. That is the reason we are not deprecation yet at phase 2.

In fact if everything go OK at some point phase 2 and phase 3 could be merge in just one phase - deprecation - removed.

Author: At some state we will remove fallback to configmap.

Reviewer 3: You need to ensure that every user has migrated from the configmap to the CR at the point you do this, have you thought at all about how you'll ensure this?

Some things you might want to consider:
* Automatic translation to the new format
* A Cluster Fleet Evaluation that allows you to inspect which clusters have the configmap but not the CR
* An upgrade block that prevents users from upgrading until they remove the configmap (this requires staff eng approval)

Author: We'll can ensure that with the deprecation procedure. This question is also in open question. 

- How do we tell cluster admins that they need to migrate? CMO conditions could be an option as well as alerting and Insights?- How do we know how many clusters have migrated to the new CRD? Telemetry?

We can choose different possibilities and different ways to do it, after the development (and the issue we will find in the process) and depends and how difficult the migration will be we can think more about that and be more specific.

Reviewer 3: > How do we tell cluster admins that they need to migrate? 

Documentation and release notes are the first step

> CMO conditions could be an option as well as alerting and Insights?

Eventually you can mark CMO as `upgradeable=false` if the configmap exists

> How do we know how many clusters have migrated to the new CRD? Telemetry?

Use [Cluster Fleet Evaluation](https://github.com/openshift/enhancements/blob/7d163511f45e73e707f8ec17f34a446772b2cff6/dev-guide/cluster-fleet-evaluation.md)

Author: Add info to PR
